gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/translator/AnonymizingTypeTranslator.java

```diff
@@ -0,0 +1,170 @@
+/*
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.apache.tinkerpop.gremlin.process.traversal.translator;
+
+import org.apache.tinkerpop.gremlin.process.traversal.*;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.apache.tinkerpop.gremlin.structure.VertexProperty;
+
+import java.sql.Timestamp;
+import java.util.*;
+
+
+/**
+ * This Translator will translate {@link Bytecode} into a representation that has been stripped of any user data
+ * (anonymized). A default anonymizer is provided, but can be replaced with a custom anonymizer as needed. The
+ * default anonymizer replaces any String, Numeric, Date, Timestamp, or UUID with a type-based token. Identical values
+ * will receive the same token (e.g. if "foo" is assigned "string0" then all occurrences of "foo" will be replaced
+ * with "string0").
+ */
+public class AnonymizingTypeTranslator extends GroovyTranslator.DefaultTypeTranslator {
+
+    /**
+     * Customizable anonymizer interface.
+     */
+    public interface Anonymizer {
+        /**
+         * Return an anonymized token for the supplied object.
+         *
+         * @param obj a {@link Traversal} object of one of the following types: String, Long, Double, FLoat, Integer,
+         *            Class, TImestamp, Date, UUID, {@link Vertex}, {@link Edge}, {@link VertexProperty}
+         * @return    an anonymized version of the supplied object
+         */
+        Object anonymize(Object obj);
+    }
+
+    /**
+     * This default implementation keeps a map from type (Java Class) to another map from instances to anonymized
+     * token.
+     */
+    public static class DefaultAnonymizer implements Anonymizer {
+        /*
+         * Map<ClassName, Map<Object, AnonymizedValue>>
+         */
+        private final Map<String, Map<Object, String>> simpleNameToObjectCache = new HashMap<>();
```

Comment:
Correct me if I am wrong here but this map will not be garbage collected until JVM is destroyed (since it is allocated on static initialization of this class). With large number of queries, this will eat up heap memory quickly. 

Is it possible to use weak reference here or clear the map or move this class outside (non-inner)?

Reply:
The map is not static, just the class. So the cache will be garbage collected when the Anonymizer goes out of scope (right after translation).

Reply:
Yes I understand that. I was wondering if there is a possibility where the class does not go out of scope through the lifecycle of JVM but seems like we create a new instance of anonymizer every time and it goes out of scope quite quickly, so my point is not valid.